frontend/src/metabase/common/components/FilterPicker/FilterPicker.unit.spec.tsx

```diff
@@ -425,34 +424,87 @@ describe("FilterPicker", () => {
     }
 
     it("should create a filter with a custom expression", async () => {
-      const { onSelect, onSelectLegacy } = setup();
+      const { query, getNextFilter, onSelect } = setup();
 
       userEvent.click(screen.getByText(/Custom expression/i));
       await editExpressionAndSubmit("[[Total] > [[Discount]{enter}");
 
       await waitFor(() =>
-        expect(onSelectLegacy).toHaveBeenCalledWith([
-          ">",
-          ["field", ORDERS.TOTAL, null],
-          ["field", ORDERS.DISCOUNT, null],
-        ]),
+        expect(onSelect).toHaveBeenCalledWith(expect.anything()),
       );
-      expect(onSelect).not.toHaveBeenCalled();
+
+      const filter = getNextFilter();
+      const { args, operator, options } = Lib.expressionParts(query, 0, filter);
+      const [left, right] = args;
+      const leftMetadata = checkNotNull(left) as Lib.ColumnMetadata;
+      const rightMetadata = checkNotNull(right) as Lib.ColumnMetadata;
+
+      expect(args).toHaveLength(2);
+      expect(Lib.displayInfo(query, 0, leftMetadata)).toMatchObject({
+        name: "TOTAL",
+        displayName: "Total",
+        semanticType: null,
+        longDisplayName: "Total",
+        effectiveType: "type/Float",
+        table: {
+          name: "ORDERS",
+          displayName: "Orders",
+          longDisplayName: "Orders",
+          isSourceTable: true,
+        },
+      });
+      expect(Lib.displayInfo(query, 0, rightMetadata)).toMatchObject({
+        name: "DISCOUNT",
+        displayName: "Discount",
+        semanticType: "type/Discount",
+        longDisplayName: "Discount",
+        effectiveType: "type/Float",
+        table: {
+          name: "ORDERS",
+          displayName: "Orders",
+          longDisplayName: "Orders",
+          isSourceTable: true,
+        },
+      });
+      expect(operator).toBe(">");
+      expect(options).toEqual({});
     });
 
     it("should update a filter with a custom expression", async () => {
-      const { onSelect, onSelectLegacy } = setup(createQueryWithCustomFilter());
+      const { query, getNextFilter, onSelect } = setup(
+        createQueryWithCustomFilter(),
+      );
 
-      await editExpressionAndSubmit("{selectall}{backspace}[[Total] > 100");
+      await editExpressionAndSubmit("{selectall}{backspace}[[Total] > 100", {
+        delay: 50,
```

Comment:
I haven't figured out why but this test will fail without the `delay` - input state won't be properly updated otherwise (the `{selectall}{backspace}` won't have an effect).